src/input/type_generic.cc

```diff
@@ -0,0 +1,116 @@
+/*
+ * type_generic.cc
+ *
+ *  Created on: May 1, 2012
+ *      Author: jb
+ */
+
+
+#include <input/type_generic.hh>
+#include <input/type_repository.hh>
+
+#include <boost/functional/hash.hpp>
+
+
+
+namespace Input {
+
+namespace Type {
+
+
+/*******************************************************************
+ * implementation of Parameter
+ */
+
+Parameter::Parameter(const string & parameter_name)
+: name_(parameter_name) {}
+
+
+Parameter::Parameter(const Parameter & other)
+: name_(other.name_) {}
+
+
+string Parameter::type_name() const {
+    return name_;
+}
+
+
+TypeBase::TypeHash Parameter::content_hash() const {
+	TypeHash seed=0;
+	boost::hash_combine(seed, "Parameter");
+	boost::hash_combine(seed, type_name());
+
+	return seed;
+}
+
+
+bool Parameter::valid_default(const string &str) const {
+    ASSERT(false, "Method valid_default can't be called for Parameter type.\n");
+    return true;
+}
+
+
+TypeBase::MakeInstanceReturnType Parameter::make_instance(std::vector<ParameterPair> vec) const {
+	for (std::vector<ParameterPair>::iterator vec_it=vec.begin(); vec_it!=vec.end(); vec_it++) {
+		if ( (*vec_it).first == this->name_ ) {
+			this->parameter_map_[(*vec_it).first] = (*vec_it).second->content_hash();
+			return std::make_pair( (*vec_it).second, this->parameter_map_ );
+		}
+	}
+	ASSERT(false, "Parameterized key '%s' in make_instance method wasn't replaced!\n", this->type_name().c_str());
```

Comment:
Imagine you get this error and you know only interface and nothing about implementation ...

I suggest to convert this assert into exception:
"No input type substitution for input type parameter "%s" found during creation of instance with parameter list: ..."
Provide parameter name through EI object.
Then catch this exception in Instance::make_instance and add EI object with list of parameter in the parameter list. This can really help to track down the error.